compiler/src/iree/compiler/Codegen/Utils/GPUUtils.cpp

```diff
@@ -556,40 +556,122 @@ Value emitGPUGroupReduction(Location loc, OpBuilder &builder, Value input,
   return laneVal;
 }
 
-std::optional<SmallVector<int64_t>> getWmmaNativeVectorSize(Operation *op) {
-  // Currently hardcode the size of wmma operation. When more cases are
-  // supported this should be picked based on what the backend supports.
-  int64_t m = 16;
-  int64_t n = 16;
+/// Helpers for unrolling to native WMMA sizes.
+static void walkSSAChain(Operation *op, WMMAFragType type,
+                         llvm::DenseMap<Operation *, WMMAFragType> &mapping) {
+  // We only want to propagate through certain ops that are used directly by the
+  // contraction.
+  if (!op ||
+      !((isa<vector::TransferWriteOp>(op) && type == WMMAFragType::C) ||
+        isa<vector::TransferReadOp>(op) ||
+        isa<vector::ExtractStridedSliceOp>(op) || isa<memref::SubViewOp>(op) ||
+        isa<memref::AllocOp>(op) || OpTrait::hasElementwiseMappableTraits(op)))
+    return;
+
+  if (mapping.contains(op)) {
+    assert(mapping[op] == type && "Remapping WMMA op in chain");
+    return;
+  }
+  mapping[op] = type;
+  for (auto operand : op->getOperands())
+    walkSSAChain(operand.getDefiningOp(), type, mapping);
+  for (auto result : op->getResults()) {
+    for (auto user : result.getUsers()) walkSSAChain(user, type, mapping);
+  }
+}
```

Comment:
Not too familiar with the slicing utilities, but from what I can tell it will be tricky to get to work for this case. The tricky part comes when unrolling on a leading fill. Essentially the fill can store to some subview and the the contract will read from a different one, the only connection being that they are both storing/reading the same memref. Thus to find that the fill is "connected" to a contract we need to slice upwards to an earliest shared memref (at a minimum the allocation/binding) and then slice downwards to the fill store.